Source/DafnyCore/AST/IndentationFormatter.cs

```diff
@@ -0,0 +1,1875 @@
+  /// That trivia will be processed with the options
+  /// * trailingTrivia == false
+  /// * precededByNewline == false
+  /// * indentationBefore == "    "
+  /// * lastIndentation == "    "
+  /// However, because there are no newlines and it's not preceded
+  /// by a newline, the method will return the same:
+  /// " "
+  /// 
+  /// 4. The leading trivia of the token `}` is
+  /// "// One last word\n"
+  /// That trivia will be processed with the options
+  /// * trailingTrivia == false
+  /// * precededByNewline == true
+  /// * indentationBefore == "  "
+  /// * lastIndentation == ""
+  /// This is an example where a comment before this token should not
+  /// be aligned with that token, the method will return:
+  /// "  // One last word\n"
+  ///
+  /// After all these formattings, the final result will be the expected one
+  /// ```
+  /// module X {
+  ///   datatype A =
+  ///     C()
+  ///     // Comment
+  ///     // continued
+  ///
+  ///   // Comment here
+  ///   const b := 1
+  ///
+  ///   // One last word
+  /// }
+  /// ```
+  public string Reindent(IToken token, bool trailingTrivia, bool precededByNewline, string indentationBefore, string lastIndentation) {
+    var input = trailingTrivia ? token.TrailingTrivia : token.LeadingTrivia;
+    var commentExtra = "";
+    // Invariant: Relative indentation inside a multi-line comment should be unchanged
+    var originalCommentIndent = 0;
+    var newCommentIndent = 0;
+    var previousMatchWasSingleLineCommentToAlign = false;
+
+    // Apply the given rules on a match of a (newline|beginning) + space + comment?
+    string ApplyIndentRules(System.Text.RegularExpressions.Match match) {
+      if (match.Groups["trailingWhitespace"].Success) {
+        return HelperString.RemoveTrailingWhitespace ? "" : match.Groups["trailingWhitespace"].Value;
+      }
+
+      var startOfString = match.Groups["previousChar"].Value == "";
+      var commentType = match.Groups["commentType"].Value;
+      if (startOfString && !precededByNewline) {
+        if (commentType.StartsWith("//")) {
+          // Possibly align consecutive // trailing comments
+          if (originalCommentIndent == 0) {
+            originalCommentIndent = token.col - 1 + token.val.Length + match.Groups["currentIndent"].Value.Length;
+            newCommentIndent = GetNewTokenVisualIndent(token, 0) + token.val.Length + match.Groups["currentIndent"].Value.Length;
+            previousMatchWasSingleLineCommentToAlign = true;
+          }
+        }
+
+        if (HelperString.RemoveTrailingWhitespace && commentType.StartsWith("\r") || commentType.StartsWith("\n")) {
+          precededByNewline = true;
+          return commentType;
+        }
+
+        if (!commentType.StartsWith("/*")) {
+          return match.Groups[0].Value;
+        }
+      }
+
+      if (commentType.Length == 0) {
+        //End of the string. Do we indent or not?
+        return precededByNewline && !trailingTrivia ? lastIndentation :
```

Comment:
Could you simplify to `trailingTrivia ? "" : (precededByNewline ? lastIndentation : match.Groups[0].Value)`

Reply:
> `trailingTrivia ? "" : (precededByNewline ? lastIndentation : match.Groups[0].Value)`

Nicer equivalent boolean formula indeed.